classes/task/process_dataflow_ad_hoc.php

```diff
@@ -34,22 +33,32 @@ class process_dataflow_ad_hoc extends \core\task\adhoc_task {
      */
     public function execute() {
         $dataflowrecord = $this->get_custom_data();
+        \tool_dataflows\task\process_dataflows::execute_dataflow($dataflowrecord->dataflowid);
+    }
+
+    /**
+     * Create an ad-hoc task for the given dataflow record
+     * and schedules for it to be run.
+     *
+     * @param object $dataflowrecord
+     * @return void
+     */
+    public static function execute_from_record($dataflowrecord): void {
+        $dataflow = new dataflow($dataflowrecord->dataflowid);
+        $task = new process_dataflow_ad_hoc();
+        $task->set_custom_data($dataflowrecord);
+
+        // For concurrent tasks, queue them up as an independant adhoc task.
+        if ($dataflow->is_concurrency_enabled()) {
+            \core\task\manager::queue_adhoc_task($task);
+            return;
+        }
 
-        try {
-            $dataflow = new dataflow($dataflowrecord->dataflowid);
-            if ($dataflow->enabled) {
-                mtrace("Running dataflow $dataflow->name as ad-hoc task (ID: $dataflowrecord->dataflowid), time due: " .
-                    userdate($dataflowrecord->nextruntime));
-                $engine = new engine($dataflow, false);
-                $engine->execute();
-                $metadata = $engine->is_blocked();
-                if ($metadata) {
-                    mtrace("Dataflow $dataflow->name locked (ID: $dataflowrecord->dataflowid). Lock data, time: " .
-                            userdate($metadata->timestamp) . ", process ID: $metadata->processid.");
-                }
-            }
-        } catch (\Throwable $thrown) {
-            mtrace("Dataflow run failed for ID: $dataflowrecord->dataflowid, " . $thrown->getMessage());
+        // Only available 3.7 onwards.
+        if (function_exists('\core\task\manager::reschedule_or_queue_adhoc_task')) {
+            \core\task\manager::reschedule_or_queue_adhoc_task($task);
+        } else {
+            \core\task\manager::queue_adhoc_task($task);
```

Comment:
Could we de-nest this else block?
```suggestion
            return;
        }

        \core\task\manager::queue_adhoc_task($task);
```